src/librustc_codegen_ssa/back/link.rs

```diff
@@ -122,11 +118,16 @@ fn link_binary_output<'a, B: ArchiveBuilder<'a>>(sess: &'a Session,
             .prefix("rmeta")
             .tempdir_in(out_filename.parent().unwrap())
             .unwrap_or_else(|err| sess.fatal(&format!("couldn't create a temp dir: {}", err)));
+        eprintln!("link_binary_output -> emit_metadata");
         let metadata = emit_metadata(sess, codegen_results, &metadata_tmpdir);
-        if let Err(e) = fs::rename(metadata, &out_filename) {
-            sess.fatal(&format!("failed to write {}: {}", out_filename.display(), e));
+        match fs::rename(&metadata, &out_filename) {
+            Ok(_) => {
+                eprintln!("METADATA WRITTEN {:?} -> {:?}", metadata, out_filename);
+                sess.parse_sess.span_diagnostic.maybe_emit_json_directive(
+                    format!("metadata file written: {:?}", out_filename));
```

Comment:
You could keep the `if let` and put this code after the closing `}` of the `if let`. Because `sess.fatal` unwinds, this will not make this directive get emitted when the rename actually failed.